### DISCUSSION TITLE
Common questions

### DISCUSSION THREAD

**Discussion by Excel-lent** [Excel-lent/ClooWrapperVBA#3] [Q&A]
@6DiegoDiego9, I mean that VBA+OpenCL performance on CPU (divided by number of CPU threads) was twice better than native C# performance (no OpenCL, single thread). For example, OpenCL gives me 8x better performance in 4 threads than native C# in a single thread on the same task. Therefore, OpenCL is twice as fast as native C# in one thread. Of course, C# compiles to CIL and OpenCL CPU/GPU driver compiles to optimized native code, but the two times difference is pretty amazing.
ClooWrapperVba is definitely not a complex library. It can set input values and arrays of integers, floats and doubles and read the outputs (arrays only), compile and run OpenCL sources (C99) and get some configuration values of available CPUs/GPUs. In this case, VBA is just a host and has nothing to do with OpenCL code that you will run, VBA only gets the results of calculations.
If you can convert your code to OpenCL (C99), you'll have a much faster version than anything you've used before.
Thank you for the link to TwinBASIC. I'll try it.
BTW I have very limited number of answers from other users and I want to know if all the CPU/GPU devices you are using have been detected?

**Comment by 6DiegoDiego9** [comment]
Thanks for the clarifications.
Unfortunately my knowledge about OpenCL and GPU programming is still *zero*.  I can only tell that yesterday I struggled a bit to make your bas+cls set to work (eg. the cls imports as modules) and finally it seemed to detect the GPU of my LG Gram 16 notebook, although I remember an "Intel" name where I expected "NVidia". I didn't pay much attention though.

I understood that OpelCL work with arrays of integers, floats and doubles. I work mostly with long and longlong, does it handle them too? and what kind of operations does it accelerate?

**Comment by Excel-lent** [comment]
So, "Long" in VBA represents "int" in C, "int64" (Longlong type in VBA) I have not implemented because it is not supproted in 32 bit Excel (but you can simply use double if you need longlong). 
OpenCL program (written in a C99, which is a "dialect" of C) is used for time-expensive math. In my case I used it to calculate covariance matrices and do some time-consuming optimization.
